### ISSUE TITLE
Nested GraphQL Versions subquery is not lazily executed

### ISSUE THREAD

**Issue by robbieaverill** [silverstripe/silverstripe-versioned#147]
**Version: 1.2.x-dev**

The nested subquery for Versions is not lazily executed, meaning that any pagination arguments are applied to the result rather than the initial query. This will potentially lead to performance issues in future:

https://github.com/silverstripe/silverstripe-versioned/blob/df62ffa2616212f4b6000ec7c7b51558e5414e61/src/GraphQL/Operations/ReadVersions.php#L50-L51

**Comment by tractorcow** [comment]
Ah yes, in 5.x we've removed this method, and recommend get_all_versions() instead. That's the lazy-equivalent, although it misses some of the "Versioned_Version" logic.

**Comment by robbieaverill** [comment]
For reference this also means that we can't pass GraphQL filters into the original DataQuery, for example "only show unpublished versions"

**Comment by chillu** [comment]
Right, that's not going to fly for 4.x. A page can have hundreds or even thousands of versions, that's why we have pagination in the first place (apart from the UX impact). So we're pushing versions to an `ArrayList`: https://github.com/silverstripe/silverstripe-versioned/blob/1/src/Versioned.php#L1926

`Versioned::get_all_versions()` exists in 4.x (or versioned 1.x) as well - can we just use that?

**Comment by robbieaverill** [comment]
## GraphQL performance issues

So the situation is that the versioned module has some GraphQL extensions with add decorations to allow queries to specify filters and columns such as `Versions` (to get a list of a versioned DataObject's historic changes).

The filter arguments are not provided to the methods like this that actually execute the DataList returning this data. Immediately obvious examples of this are pagination and filters like "only show unpublished versions." Pagination _does_ work in history viewer (using this GraphQL API), but my suspicion is that the `->Versions()` DataList loads the entire history then applies the pagination edges before returning it to the client - will be an issue with a moderate amount of records.

## Versioned API problems

Next in the line is the extensibility of the Versioned_Version API which is causing problems all over the CWP 2 ecosystem because it assembles a custom SQL query rather than using a DataList. Some quick examples:

* tractorcow/silverstripe-fluent can't filter the list of versions by the current locale (https://github.com/tractorcow/silverstripe-fluent/issues/414)
* silverstripe/versionfeed doesn't work with fluent (https://github.com/silverstripe/silverstripe-versionfeed/issues/55)
* silverstripe/versioned-admin (history viewer) can't add "toggle unpublished versions" controls (#105)

Note that @tractorcow has rewritten this API in a breaking way for SilverStripe 5.x and removed the use of Versioned_Version and now uses `Versioned:: get_all_versions()` which returns a DataList. This API also exists in SS4, but we would need to refactor any touch points in the SilverStripe 4.x line that currently use the Versioned_Version API to use a DataList instead. Examples of this are the CMSPageHistoryController in the CMS and the new GraphQL extensions like this: https://github.com/silverstripe/silverstripe-versioned/blob/df62ffa2616212f4b6000ec7c7b51558e5414e61/src/GraphQL/Operations/ReadVersions.php#L50-L51

**Comment by robbieaverill** [comment]
This would be fixed by #182

**Comment by robbieaverill** [comment]
Status update on this ticket:

The original problem is that the GraphQL query uses Versioned::Versions() which returns an ArrayList of Versioned_Version instances. This isn't lazily executed like a DataList, so suffer on performance in large data sets.

The PR at #182 backports functionality from SS5 that exposes a similar construct as a DataList, and switches the GraphQL query to return that instead. It's no longer a list of Versioned_Version instances, but I've left the original method in place to preserve backwards compatibility. Technically the return type of the GraphQL class has changed from ArrayList to DataList, but since it's expected to be used via a GraphQL query the end result hasn't changed so I kind of think it's an OK semver break. I've updated the tests to look for an SS_List instead of an ArrayList so it'd cover both.
